engine/access/rest/blocks.go

```diff
@@ -194,31 +110,77 @@ func getBlockPayloadByID(req *requestDecorator, backend access.API, _ LinkGenera
 		return nil, NewBadRequestError(err.Error(), err)
 	}
 
-	_, payload, statusErr := blockLookup(req.Context(), id, backend)
-	if err != nil {
+	blkProvider := NewBlockProvider(backend, withID(&id))
+	blk, statusErr := blkProvider.getBlock(req.Context())
+	if statusErr != nil {
 		return nil, statusErr
 	}
-
+	payload := blockPayloadResponse(blk.Payload)
 	return payload, nil
 }
 
-func blockLookup(ctx context.Context, id flow.Identifier, backend access.API) (*generated.BlockHeader, *generated.BlockPayload, StatusError) {
-	flowBlock, err := backend.GetBlockByID(ctx, id)
-	if err != nil {
-		return nil, nil, idLookupError(id, "block", err)
+func getBlock(blkProvider *blockProvider, req *requestDecorator, backend access.API, link LinkGenerator) (*generated.Block, StatusError) {
+	var responseBlock = new(generated.Block)
+	responseBlock.Expandable = new(generated.BlockExpandable)
+
+	var id flow.Identifier
+	// if payload is to be expanded then lookup full block which contains both header and payload
+	if req.expands(ExpandableFieldPayload) {
+		blk, err := blkProvider.getBlock(req.Context())
+		if err != nil {
+			return nil, err
+		}
+		responseBlock.Header, responseBlock.Payload = blockHeaderResponse(blk.Header), blockPayloadResponse(blk.Payload)
+		id = blk.ID()
+	} else {
+
+		// else only lookup header and add expandable link for payload
+		header, statusErr := blkProvider.getHeader(req.Context())
+		if statusErr != nil {
+			return nil, statusErr
+		}
+		responseBlock.Header = blockHeaderResponse(header)
+		responseBlock.Payload = nil
```

Comment:
assigning `nil` is not necessary, this is the default value.

Reply:
yeah i did that just so that the reader knows that we sent a nil payload